my_variant.h

```diff
@@ -0,0 +1,118 @@
+#include <cassert>
+#include <iostream>
+#include <algorithm>
+#include <type_traits>
+
+using std::cout;
+using std::endl;
+
+
+template <typename... types>
+struct max_size_helper { };
+
+template <typename T, typename... types>
+struct max_size_helper<T, types...> {
+	static size_t constexpr value = std::max(sizeof(T), max_size_helper<types...>::value);
+};
+
+template <> 
+struct max_size_helper<> {
+	static size_t constexpr value = 0;
+};
+
+
+template <typename... types>
+struct max_align_helper { };
+
+template <typename T, typename... types>
+struct max_align_helper<T, types...> {
+	static size_t constexpr value = std::max(alignof(T), max_align_helper<types...>::value);
+};
+
+template <> 
+struct max_align_helper<> {
+	static size_t constexpr value = 0;
+};
+
+/**
+* get index of type from parameter pack
+*/
+
+template <typename T, typename... Ts> struct get_index;
+
+template <typename T, typename... Ts>
+struct get_index<T, T, Ts...> : std::integral_constant<std::size_t, 0> {};
+
+template <typename T, typename Tail, typename... Ts>
+struct get_index<T, Tail, Ts...> :
+    std::integral_constant<std::size_t, 1 + get_index<T, Ts...>::value> {};
+
+template <typename T>
+struct get_index<T> {
+	static_assert(sizeof(T) == 0, "Element is not found");
+};
+
+template <class Obj, class... types> struct destroy_all;
+template <class T, class... types>
+struct my_variant {
+	my_variant() : m_member(new (&m_aligned_storage) T{}) { }
+
+	my_variant(my_variant<T, types...>&) {
+		assert(sizeof(T) == 0 && "Not implemented");
+	}
+	my_variant(my_variant<T, types...>&&) {
+		assert(sizeof(T) == 0 && "Not implemented");
+	}
+	// need to add emplacement ctors...
+
+	template <class U>
+	my_variant(U&& u) :
+		m_index(get_index<U, T, types...>::value),	
+ 		m_member(new (&m_aligned_storage) U{u} )
+	{ }
+	~my_variant() {
+		destroy_all<decltype(*this), T, types...>::f(*this);	
```

Comment:
You solved the destructor problem "ad hoc", which is fine :). But actually, it was just a symptom of the more general problem: what if you (or the user!) had to apply some other function to the contained value (e.g. print() etc)? In your case the only such function was the destructor, that's why you encountered this problem here.
So, the general problem is the following: we need a *generic* mechanism that would allow us to apply any function to the *contained* type. When you have such a mechanism, you're able to invoke it in the destructor as well! (Hint: usually we call such a mechanism "visitor".)

Reply:
I added a visitor.